rust/src/normalization.rs

```diff
@@ -120,6 +121,294 @@ pub(crate) fn rewrite_rpmdb_timestamps<F: Read + Write + Seek>(rpmdb: &mut F) ->
+    }
+}
+
+mod rpmdb_bdb {
+    use anyhow::{anyhow, Context, Result};
+    use binread::{BinRead, BinReaderExt};
+    use lazy_static::lazy_static;
+    use openat::SimpleType;
+    use sha2::Digest;
+    use std::ffi::OsStr;
+    use std::io::{Read, Seek, SeekFrom, Write};
+    use std::os::unix::io::AsRawFd;
+    use std::path::{Path, PathBuf};
+    use subprocess::{Exec, Redirection};
+
+    #[derive(BinRead, Debug, Clone, Copy, PartialEq, Eq)]
+    #[repr(u8)]
+    #[br(repr=u8)]
+    enum PageType {
+        IBTree = 3,
+        LBTree = 5,
+        Overflow = 7,
+        HashMeta = 8,
+        BTreeMeta = 9,
+        Hash = 13,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct MetaHeader {
+        lsn: u64,
+        pgno: u32,
+        magic: u32,
+        version: u32,
+        pagesize: u32,
+        encrypt_alg: u8,
+        page_type: PageType,
+        metaflags: u8,
+        unused1: u8,
+        free: u32,
+        last_pgno: u32,
+        nparts: u32,
+        key_count: u32,
+        record_count: u32,
+        flags: u32,
+        uid: [u8; 20],
+    }
+
+    const BTREE_MAGIC: u32 = 0x00053162;
+    const HASH_MAGIC: u32 = 0x00061561;
+    const PAGE_HEADER_SIZE: u16 = 26;
+    const PAGE_HEADER_MAGIC_OFFSET: u64 = 0x34;
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct PageHeader {
+        lsn: u64,
+        pgno: u32,
+        prev_pgno: u32,
+        next_pgno: u32,
+        entries: u16,
+        hf_offset: u16,
+        level: u8,
+        page_type: PageType,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(repr=u8)]
+    #[repr(u8)]
+    enum BTreeItemType {
+        KeyData = 1,
+        Duplicate = 2,
+        Overflow = 3,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct BTreeItem {
+        len: u16,
+        item_type: BTreeItemType,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(repr=u8)]
+    #[repr(u8)]
+    enum HashItemType {
+        KeyData = 1,
+        Duplicate = 2,
+        Offpage = 3,
+        OffDup = 4,
+    }
+
+    lazy_static! {
+        static ref PROC_SELF_CWD: PathBuf = PathBuf::from("/proc/self/cwd");
+        static ref PROC_SELF_FD: PathBuf = PathBuf::from("/proc/self/fd");
+    }
+
+    pub(super) fn normalize_rpmdb_bdb<P: AsRef<Path>>(
+        rootfs: &openat::Dir,
+        rpmdb_path: P,
+        timestamp: u32,
+    ) -> Result<()> {
+        let rpmdb_path = rpmdb_path.as_ref();
+
+        for entry in rootfs.list_dir(rpmdb_path)? {
+            let entry = entry?;
+
+            match entry.simple_type() {
+                Some(SimpleType::File) => (),
+                _ => continue,
+            };
+
+            if let Some(filename) = entry.file_name().to_str() {
+                if filename.starts_with('.') || filename.starts_with("__db") {
+                    continue;
+                }
+            } else {
+                continue;
+            }
```

Comment:
Optional, but I like using combinators:
```suggestion
            if filename.filter(|n| !(n.starts_with('.') || n.starts_with("__db"))).is_some() {
                continue
            }
```

Reply:
Does this look ok?

```rust
            if entry
                .file_name()
                .to_str()
                .filter(|name| !(name.starts_with('.') || name.starts_with("__db")))
                .is_none()
            {
                continue;
            }
```